### PR TITLE
[dev] Support per-endpoint expose parameters in juju expose/unexpose CLI commands

### PR DESCRIPTION
**Pull request by achilleasa**
## Description of change

This PR updates the `juju expose` and `juju unexpose` commands work with per-endpoint expose settings. More specificically:

- The `juju expose` command now supports the optional flags: --endpoints, --to-spaces, --to-cidrs. If none is specified, the command behaves like a typical `juju expose <application>` invocation as you would find in a pre 2.9 client/controller combo.
- The `juju unexpose` command supports the optional flag --endpoints. If not specified, the command behaves like a typical `juju unexpose <application>` invocation as you would find in a pre 2.9 client/controller combo.

A key change in the behavior of these two commands is that they work in an additive way: you can run multiple invocations and the rules accumulate. However, note that any rules that apply to a particular endpoint overwrite any pre-existing rules for the same endpoint.

To support this new flags, the payload for the Expose/Unexpose API calls had to be augmented with an extra field. The calls themselves are backwards compatible with older clients and the Application facade has already been bumped on the develop branch in another PR so no further bump was required.

The client side performs facade version checking and raises errors when trying to pass the new endpoint-specific parameters to an older controller. This prevents a potential security hole where the operator specifies endpoint-specific parameters that are ignored by an old controller thus causing **all ports** of the application to become accessible from 0.0.0.0/0!

_NOTE: these changes just manipulate the application document. The firewall worker still only considers the `exposed` flag when generating ingress rules. Currently, the only way to see the results of these changes is to query the DB._

To make reviewing easier, the following chunks of pending work will be addressed via separate PRs:
- juju bundle-diff
- juju export-bundle support for endpoint-specific parameters.
- juju show-application support for endpoint-specific parameters.
- integration tests

#### Single charm

```console
$ juju deploy lxd test-29 --no-gui
$ juju deploy apache2
$ juju run --unit apache2/0 "open-port 80/tcp"
$ juju run --unit apache2/0 "open-port 1337/tcp --endpoints logs"

# Expose everything and then override the expose settings for two endpoints
$ juju expose apache2
$ juju expose apache2 --endpoints logs,website --to-cidrs 10.0.0.0/24

# Check mongo
juju:PRIMARY> db.applications.find({}, {"exposed":1,"exposed-endpoints":1}).pretty()
{
        "_id" : "fb2c5a6a-26c9-45f3-8c5b-dccbc87f393c:apache2",
        "exposed" : true,
        "exposed-endpoints" : {
                "logs" : {
                        "to-cidrs" : [
                                "10.0.0.0/24"
                        ]
                },
                "website" : {
                        "to-cidrs" : [
                                "10.0.0.0/24"
                        ]
                }
        }
}

# Override existing rules for endpoint
$ juju expose apache2 --endpoints website --to-spaces alpha
juju:PRIMARY> db.applications.find({}, {"exposed":1,"exposed-endpoints":1}).pretty()
{
        "_id" : "fb2c5a6a-26c9-45f3-8c5b-dccbc87f393c:apache2",
        "exposed" : true,
        "exposed-endpoints" : {
                "logs" : {
                        "to-cidrs" : [
                                "10.0.0.0/24"
                        ]
                },
                "website" : {
                        "to-spaces" : [
                                1
                        ]
                }
        }
}

# Unexpose one of the endpoints. The application should remain exposed
$ juju unexpose --endpoint logs

juju:PRIMARY> db.applications.find({}, {"exposed":1,"exposed-endpoints":1}).pretty()
{
        "_id" : "fb2c5a6a-26c9-45f3-8c5b-dccbc87f393c:apache2",
        "exposed" : true,
        "exposed-endpoints" : {
                "website" : {
                        "to-spaces" : [
                                1
                        ]
                }
        }
}

# Unexpose second endpoint. The application should now become unexposed
$ juju unexpose --endpoint website

juju:PRIMARY> db.applications.find({}, {"exposed":1,"exposed-endpoints":1}).pretty()
{
        "_id" : "fb2c5a6a-26c9-45f3-8c5b-dccbc87f393c:apache2",
        "exposed" : false,
        "exposed-endpoints" : {}
}
```

#### Bundle

```console
$ juju deploy lxd test-29

# Create the following bundle
$ cat bundle1.yaml
applications:
  apache2:
    charm: apache2
    expose: true
--- #overlay
applications:
  apache2:
    charm: apache2
    exposed-endpoints:
      logs:
        expose-to-cidrs: ["0.0.0.0/0"]

$ juju deploy ./bundle1.yaml
ERROR cannot deploy bundle: the provided bundle has the following errors:
exposed-endpoints cannot be specified together with "exposed:true" in application "apache2" as this poses a security risk when deploying bundles to older controllers

# Create the following bundle
$ cat bundle2.yaml
applications:
  apache2:
    charm: apache2
--- #overlay
applications:
  apache2:
    charm: apache2
    exposed-endpoints:
      website:
        expose-to-spaces: ["alpha"]
        expose-to-cidrs: ["0.0.0.0/0"]
      logs:
        expose-to-cidrs: ["10.0.0.0/24"]
      local-monitors:
        expose-to-cidrs: ["10.0.0.0/24"]

$ juju deploy ./bundle2.yaml
Resolving charm: apache2
Executing changes:
- upload charm cs:apache2-35
- deploy application apache2 using cs:apache2-35
- override expose settings for endpoints local-monitors,logs of apache2 and allow access from CIDR 10.0.0.0/24
- override expose settings for endpoint website of apache2 and allow access from space alpha and CIDR 0.0.0.0/0
Deploy of bundle completed.

juju:PRIMARY> db.applications.find({}, {"exposed":1,"exposed-endpoints":1}).pretty()
{
        "_id" : "fb2c5a6a-26c9-45f3-8c5b-dccbc87f393c:apache2",
        "exposed" : true,
        "exposed-endpoints" : {
                "local-monitors" : {
                        "to-cidrs" : [
                                "10.0.0.0/24"
                        ]
                },
                "logs" : {
                        "to-cidrs" : [
                                "10.0.0.0/24"
                        ]
                },
                "website" : {
                        "to-space-ids" : [
                                "0"
                        ],
                        "to-cidrs" : [
                                "0.0.0.0/0"
                        ]
                }
        }
}
```

#### Single charm

```console
# Bootstrap a 2.8 controller from snap and use the 2.9 client from this PR
$ /snap/bin/juju bootstrap lxd test-28 --no-gui
$ juju deploy apache2
$ juju expose apache2 --endpoints logs
ERROR controller does not support granular expose parameters; applying this change would make all open application ports accessible from 0.0.0.0/0

# Running the command without the --endpoints option works as expected
$ juju expose apache2
```

#### Bundle

```console
# Bootstrap a 2.8 controller from snap and use the 2.9 client from this PR
$ /snap/bin/juju bootstrap lxd test-28 --no-gui

# Create the following bundle
$ cat bundle2.yaml
applications:
  apache2:
    charm: apache2
--- #overlay
applications:
  apache2:
    charm: apache2
    exposed-endpoints:
      website:
        expose-to-spaces: ["alpha"]
        expose-to-cidrs: ["0.0.0.0/0"]
      logs:
        expose-to-cidrs: ["10.0.0.0/24"]
      local-monitors:
        expose-to-cidrs: ["10.0.0.0/24"]

$ juju deploy ./bundle2.yaml
juju deploy ./bundle2.yaml
Resolving charm: apache2
Executing changes:
- upload charm cs:apache2-35
- deploy application apache2 using cs:apache2-35
- override expose settings for endpoints local-monitors,logs of apache2 and allow access from CIDR 10.0.0.0/24
- override expose settings for endpoint website of apache2 and allow access from space alpha and CIDR 0.0.0.0/0
ERROR cannot deploy bundle: cannot expose application apache2: controller does not support granular expose parameters; applying this change would make all open application ports accessible from 0.0.0.0/0
```

## Documentation changes

The command help texts for expose/unexpose have been updated but we still need to update the juju docs.

### CODE REVIEW

File: apiserver/facades/client/application/application_test.go [hunk 1]
```diff
@@ -2493,53 +2532,108 @@ func (s *applicationSuite) setupApplicationExpose(c *gc.C) {
...
+	},
+	{
+		about:       "unknown space name in expose parameters",
+		application: "dummy-application",
+		exposedEndpointParams: map[string]params.ExposedEndpoint{
+			"": {
+				ExposeToSpaces: []string{"invaders"},
+			},
+		},
+		expErr: `space "invaders" not found`,
```
Comment (achilleasa) [inline apiserver/facades/client/application/application_test.go:2590]: It's good to know that all that R&D effort for modulating shield frequency harmonics has finally paid off ;-) 👾

Comment (SimonRichardson) [reply apiserver/facades/client/application/application_test.go:2590]: :space_invader: :gun:

File: cmd/juju/application/unexpose.go [hunk 2]
```diff
@@ -19,10 +20,27 @@ Removes public availability over the network for an application.`[1:]
 var usageUnexposeDetails = `
 Adjusts the firewall rules and any relevant security mechanisms of the
 cloud to deny public access to the application.
-An application is unexposed by default when it gets created.
 
-Examples:
-    juju unexpose wordpress
+Applications are unexposed by default when they gets created. If exposed via
```
Comment (SimonRichardson) [inline cmd/juju/application/unexpose.go:24]:
```suggestion
Applications are unexposed by default when they get created. If exposed via
```

**Review by SimonRichardson — COMMENTED** [review]
Code looks good, need to Q&A it now.

**Review by SimonRichardson — APPROVED** [review]
Q&A done :+1:

**Comment by achilleasa** [comment]
!!build!!

**Comment by achilleasa** [comment]
$$merge$$ (windows test box is acting up)

**Pull request merged**